Simplenote/src/androidTest/java/com/automattic/simplenote/smoke/test/TestRunner.java

```diff
@@ -0,0 +1,405 @@
+package com.automattic.simplenote.smoke.test;
+
+import androidx.test.rule.ActivityTestRule;
+
+import com.automattic.simplenote.NotesActivity;
+import com.automattic.simplenote.smoke.data.DataProvider;
+import com.automattic.simplenote.smoke.data.NoteDTO;
+import com.automattic.simplenote.smoke.data.SignUpDataProvider;
+import com.automattic.simplenote.smoke.pages.IntroPage;
+import com.automattic.simplenote.smoke.pages.LoginPage;
+import com.automattic.simplenote.smoke.pages.MainPage;
+import com.automattic.simplenote.smoke.pages.SettingsPage.SortOrder;
+import com.automattic.simplenote.smoke.utils.TestUtils;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class TestRunner {
+    @Rule
+    public ActivityTestRule<NotesActivity> mActivityTestRule = new ActivityTestRule<>(NotesActivity.class, true, true);
+
+    private static String email;
+    private static String password;
+
+    @Before
+    public void init() {
+        TestUtils.logoutIfNecessary(DataProvider.UNKNOWN_EMAIL);
+
+        if (email == null || password == null) {
+
+            email = SignUpDataProvider.generateEmail();
+            password = SignUpDataProvider.generatePassword();
+
+            System.out.println("Email: " + email);
+            System.out.println("Password: " + password);
+
+            new IntroPage()
+                    .openSignUp()
+                    .signUp(email, password)
+                    .logout(email);
+        }
+    }
+
+    @After
+    public void tearDown() {
+    }
+
+    @Test
+    public void testLogin() {
+
+        new IntroPage()
+                .goToLoginWithEmail()
+                .login(email, password);
+
+        new MainPage()
+                .isOpened()
+                .logout(email);
+    }
+
+    @Test
+    public void testLogout() {
+
+        if (!TestUtils.logoutIfNecessary(email)) {
+            new IntroPage()
+                    .goToLoginWithEmail()
+                    .login(email, password);
+
+            new MainPage()
+                    .isOpened()
+                    .logout(email);
+        }
+    }
+
+    @Test
+    public void testLoginWithWrongPassword() {
+        TestUtils.logoutIfNecessary(email);
+
+        new IntroPage()
+                .goToLoginWithEmail()
+                .login(DataProvider.LOGIN_EMAIL, DataProvider.LOGIN_WRONG_PASSWORD);
+
+        new LoginPage()
+                .checkLoginFailedMessage();
+    }
+
+    @Test
+    public void testSearchingInTheSearchFieldDoesAGlobalSearch() {
+        NoteDTO noteDTO = DataProvider.generateNotes(1).get(0);
+
+        String searchParam = noteDTO.getTitle();
+
+        new IntroPage()
+                .goToLoginWithEmail()
+                .login(email, password);
```

Comment:
Given that the rest of the tests need to start logged in, there's a lot of time spent signing up, logging out, and logging back in here. What do you think of separating the tests into different classes that have common setup needs, to save time and avoid repetition?

Reply:
You are definitely right about it! I was thinking about how to handle it. I will apply the best possible solution for this.